autobahn/wamp/protocol.py

```diff
@@ -490,6 +490,32 @@ def _swallow_error(self, fail, msg):
             )
         return None
 
+    def type_check(self, func, return_error=False):
+        """
+        Does parameter type checking and validation against type hints
+        and appropriately tells the user code and the caller (through router).
+
+        :return_error: Whether the error should be returned to the caller in
+                       addition to being notified to the user code.
+        :type: bool
+        """
+        async def _type_check(*args, **kwargs):
+            # Converge both args and kwargs into a dictionary
+            arguments = inspect.getcallargs(func, *args, **kwargs)
+            response = []
+            for name, kind in func.__annotations__.items():
+                if name in arguments and not isinstance(arguments[name], kind):
```

Comment:
If `func` has no annotations at all, `__annotations__` is just empty, right? So that brings up a sort of meta-question, then: if you've asked for type-checking, but the procedure you've registered doesn't have any annotations should that be a `ValueError` or similar? (That is, an error because you're using the API improperly?)

That is, maybe `type_check` itself should check if `func.__annotations__` has any entries and throw a `ValueError` if it doesn't?

Reply:
I am fine either way. What do you think @oberstet ?

Reply:
my view:

if I added a type-check annotator to an endpoint procedure (for callee) or event handler I want _all_ parameters of the callable to be type checked. if I missed a type annotation for one or more parameters, the type-check annotator should throw an error.

one can argue for that error to be a ValueError or TypeError - depending on perspective (a Value is just a Type-0 type .. blabla) .. I don't care which one;)

Reply:
Opened https://github.com/crossbario/autobahn-python/issues/1317 for both suggestions above